### PR TITLE
Issue #2448. Download WFS data (KML, SHP, CSV, GML, etc.)

### LINKED ISSUE #2448: Download WFS data (KML, SHP, CSV, GML, etc.)
### Description
If for a WMS layer a corresponding WFS is available the user shall be able to download the WFS data of this layer from within the map/legend. The formats available are configured in the WFS definition (GetCapabilities).

### Please indicate if this issue is related to a bug or a new feature request
- [ ] Bug
- [X] New Feature

### In case of New Feature (otherwise remove this paragraph)

*Acceptance Criteria - AC*

We will:
 - Read the list of supported formats from WFS capabilities
 - Extend format selection of from data export of feature grid
 - Allow direct download from the TOC with a custom tool
It would be important to have a list of WFS that we intend to use in order to start testing interoperability with them.

### PR DESCRIPTION
**Pull request by mricca**
## Description
If for a WMS layer a corresponding WFS is available the user shall be able to download the WFS data of this layer from within the map/legend. The formats available are configured in the WFS definition (GetCapabilities).

## Issues
 - Fix #2448 

**Please check if the PR fulfills these requirements**
- [X] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)

 - [X] Feature

**What is the current behavior?** (You can also link to an open issue here)
The export data format from the feature grid is hard coded

**What is the new behavior?**
Now, the export data format from the feature grid is dinamically fetched by WFS getCapabilities
The formats options is still available. If configured, formats are not fetched by server
```
{
 "name": "WFSDownload",
 "cfg": {
   "formats": [
           {"name": "csv", "label": "csv"},
           {"name": "shape-zip", "label": "shape-zip"}
   ]
 }
}
```

**Does this PR introduce a breaking change?** (check one with "x", remove the other)

 - [] Yes
 - [X] No

### CODE REVIEW

File: web/client/epics/wfsdownload.js [hunk 1]
```diff
@@ -1,12 +1,42 @@
...
+	"OGR-FileGDB": "OGR-FileGDB",
+	"OGR-GPKG": "OGR-GPKG",
+	"OGR-KML": "OGR-KML",
+	"OGR-MIF": "OGR-MIF",
+	"OGR-TAB": "OGR-TAB",
+	"SHAPE-ZIP": "Shapefile",
+	"gml32": "GML3.2",
+	"application/json": "GeoJSON",
+	"csv": "CSV",
+	"application/x-gpkg": "application/x-gpkg",
```
Comment (mbarto) [inline web/client/epics/wfsdownload.js]: This one should be geopackage too

File: web/client/epics/wfsdownload.js [hunk 2]
```diff
@@ -1,12 +1,42 @@
...
 const FilterUtils = require('../utils/FilterUtils');
 const {getByOutputFormat} = require('../utils/FileFormatUtils');
+const {getLayerWFSCapabilities} = require('../observables/wfs');
+
+const DOWNLOAD_FORMATS_LOOKUP = {
+	"gml3": "GML3.1",
+	"GML2": "GML2",
+	"application/vnd.google-earth.kml+xml": "KML",
+	"OGR-CSV": "OGR-CSV",
+	"OGR-FileGDB": "OGR-FileGDB",
```
Comment (mbarto) [inline web/client/epics/wfsdownload.js]: OGR-GeoDatabase

File: web/client/epics/wfsdownload.js [hunk 3]
```diff
@@ -1,12 +1,42 @@
...
+	"OGR-GPKG": "OGR-GPKG",
```
Comment (mbarto) [inline web/client/epics/wfsdownload.js]: OGR-GeoPackage

File: web/client/components/data/download/DownloadDialog.jsx [hunk 4]
```diff
@@ -20,10 +20,13 @@ class DownloadDialog extends React.Component {
...
         onExport: PropTypes.func,
         onDownloadOptionChange: PropTypes.func,
+        onFormatOptionsFetch: PropTypes.func,
         downloadOptions: PropTypes.object,
         formats: PropTypes.array,
         srsList: PropTypes.array,
-        defaultSrs: PropTypes.string
+        defaultSrs: PropTypes.string,
+        layer: PropTypes.object,
+        formatsLoading: PropTypes.bool
```
Comment (mbarto) [inline web/client/components/data/download/DownloadDialog.jsx]: Can we add an optional filter to limit the available formats if we don't want to expose all of them? It should be configurable from the plugin .

Comment (mricca) [reply web/client/components/data/download/DownloadDialog.jsx]: Config Example:
```
            {
              "name": "WFSDownload",
              "cfg": {
                "formatFilter":  {"SHAPE-ZIP": "Shapefile"}
              }
            }
```

File: web/client/epics/wfsdownload.js [hunk 5]
```diff
@@ -14,27 +14,24 @@ const DOWNLOAD_FORMATS_LOOKUP = {
 	"GML2": "GML2",
 	"application/vnd.google-earth.kml+xml": "KML",
 	"OGR-CSV": "OGR-CSV",
-	"OGR-FileGDB": "OGR-FileGDB",
-	"OGR-GPKG": "OGR-GPKG",
+	"OGR-FileGDB": "OGR-GeoDatabase",
+	"OGR-GPKG": "OGR-OGR-GeoPackage",
```
Comment (mbarto) [inline web/client/epics/wfsdownload.js]: two times OGR?

File: web/client/epics/wfsdownload.js [hunk 6]
```diff
@@ -14,27 +14,24 @@ const DOWNLOAD_FORMATS_LOOKUP = {
...
 };
 
-const hasOutputFormat = (data) => {
+const hasOutputFormat = (data, formatFilter) => {
     const operation = get(data, "WFS_Capabilities.OperationsMetadata.Operation");
     const getFeature = find(operation, function(o) { return o.name === 'GetFeature'; });
     const parameter = get(getFeature, "Parameter");
     const outputFormatValue = find(parameter, function(o) { return o.name === 'outputFormat'; }).Value;
-    const pickedObj = pick(DOWNLOAD_FORMATS_LOOKUP, outputFormatValue);
+    const pickedObj = pick(!isEmpty(formatFilter) ? formatFilter : DOWNLOAD_FORMATS_LOOKUP, outputFormatValue);
```
Comment (mbarto) [inline web/client/epics/wfsdownload.js]: So formatFilter is an alternative to DOWNLOAD_FORMATS_LOOKUP ?

Comment (mricca) [reply web/client/epics/wfsdownload.js]: yes, it's ok?

File: web/client/epics/wfsdownload.js (hunk 5 above)
Comment (mricca) [reply web/client/epics/wfsdownload.js]: mistake

File: web/client/epics/wfsdownload.js (hunk 6 above)
Comment (mbarto) [reply web/client/epics/wfsdownload.js]: acceptable

File: web/client/components/data/download/DownloadDialog.jsx (hunk 4 above)
Comment (offtherailz) [reply web/client/components/data/download/DownloadDialog.jsx]: please add documentation for these configuration options and how to use them in the plugin's doc.

Comment (offtherailz) [reply web/client/components/data/download/DownloadDialog.jsx]: Indicate also if there are breaking changes in configuration (in the P.R.) . E.g. if you configured some formats, you have to change your configurations this way.

**Comment by offtherailz** [comment]
One more question before approval. Is there a way to skipt capabilities and use a fixed list as it was before? 
@mbarto should we keep that possibility?

**Comment by offtherailz** [comment]
If I understand well, if formats is defined, it will use the "formats". Otherwise it will fetch and filter outputFormats with the supported one, using the `DOWNLOAD_FORMATS_LOOKUP` map. 
If so, the following configuration should work without any changes to the code 
```
"formats": [
{"name": "excel", "label": "excel"},
{"name": "excel2007", "label": "excel2007"},
{"name": "dxf-zip", "label": "dxf-zip"}
]
```
Isn't it? If yes, this is ok for me too.

**Comment by mbarto** [comment]
@offtherailz It should work that way you are describing now

**Pull request merged**